generate_buddy_headers.py

```diff
@@ -0,0 +1,160 @@
+            assert err_class in err_class_mapping, f"Unknown error class {err_class} in error code {code}."
+
+            err_id = f"{err_class_mapping[err_class]}_{err['id']}"
+            assert err_id not in err_id_cache, f"Duplicate error id \"{err_id}\"."
+            err_id_cache.add(err_id)
+
+            err_code = int(code)
+            assert err_code not in err_dict, f"Duplicate error code {code}."
+
+            if mmu:
+                btns = [f"ButtonOperations::{action}" for action in err["action"]]
+
+                # Make sure the array always has 3 buttons, NoOperation button
+                # occupying the appropriate "empty" slots
+                if len(btns) == 0:
+                    btns.append("ButtonOperations::NoOperation")
+                if len(btns) == 1:
+                    btns.insert(0, "ButtonOperations::NoOperation")
+                if len(btns) == 2:
+                    btns.append("ButtonOperations::NoOperation")
+
+                btns_text = f",\n                    {{{', '.join(btns)}}}"
+            else:
+                btns_text = ""
+
+            err_dict[err_code] = {
+                "id": err_id,
+                "code": err_code,
+                "title": err["title"],
+                "text": err["text"],
+                "btns": btns_text
+            }
+
+    os.makedirs(header_file_name.parent, exist_ok=True)
+
+    enum_items = ",\n                ".join(f"{err['id']} = {err['code']}"
+                                        for err in err_dict.values())
+
+    list_items = ",".join(f"""\n                {{
+                    N_("{err['title']}"),
+                    N_("{err['text']}"),
+                    {err['id']}{err['btns']}
+                }}""" for err in err_dict.values())
+
+    if mmu:
+        content = f"""
+            #pragma once
+            #include "inttypes.h"
+            #include "i18n.h"
+            #include "button_operations.h"
+
+
+            namespace MMU2 {{
+
+            inline constexpr uint8_t ERR_MMU_CODE = {printer_code};
+
+            enum ErrorNumber : uint16_t {{
+                ERR_UNDEF = 0,
+                {enum_items}
+            }};
+
+            struct MMUErrorDesc {{
+                // 32 bit
+                const char *err_title;
+                const char *err_text;
+                // 16 bit
+                ErrorNumber err_num;
+                std::array<ButtonOperations, 3> buttons;
+            }};
+
+            inline constexpr MMUErrorDesc error_list[] = {{{list_items}
+            }};
+
+            }}  // namespace MMU2
+            """
+    else:
+        content = f"""
+            #pragma once
+            #include "inttypes.h"
+            #include "i18n.h"
+
+
+            inline constexpr uint8_t ERR_PRINTER_CODE = {printer_code};
+
+            typedef enum : uint16_t {{
+                ERR_UNDEF = 0,
+                {enum_items}
+            }} ErrorNumber;
+
+            typedef struct {{
```

Comment:
Is there a reason why ErrorDesc and ErrorNumber aren't 'normal' c++ structs? Usage of `constexpr` implies c++ therefore i'd expect it to be the same(-ish) as with mmu

Reply:
I assumed the header is used in C somewhere and haven't realized C doesn't have `constexpr`. I haven't dug too deep in this regard and mostly kept what's in `errors_list.h` now. I can change it though (and use the `.hpp` suffix for the header).